### PR TITLE
Custom JWT Token Issuer Sample

### PR DESCRIPTION
**Pull request by RajeshTammali**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: oauth2/custom-token-issuer/src/README.MD [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+Scope format in string as follows :
+"scope": "claims.read claims.write customerProfile.read customerProfile.write customerServices.read customerServices.write healthCareProviders.read healthCareProviders.write notifications.read policyAdmin.read"
+
+Sample Maven Project :
+
+Create a  maven project with custom class extends JWTTokenIssuer and Override the createJWTClaimSet method.
+Add the dependencies in pom.xml
+Build the claimset by using the JWTClaimsSet.Builder instance
+Build the maven project “mvn clean install”
+Copy the jar from target folder into  $IS_HOME/repository/components/lib-location
```
Comment (mefarazath) [inline oauth2/custom-token-issuer/src/README.MD:13]:
```suggestion
Copy the jar from target folder into  $IS_HOME/repository/components/lib location

```

File: oauth2/pom.xml [hunk 2]
```diff
@@ -1,37 +1,38 @@
...
+    <packaging>pom</packaging>
+    <name>Identity Server : OAuth2.0 Samples with Identity Server 5.2.0</name>
+
+    <modules>
+        <module>playground2</module>
+        <module>custom-grant</module>
+        <module>custom-token-issuer</module>
+    </modules>
+
+</project>
```
Comment (mefarazath) [inline oauth2/pom.xml:38]:
```suggestion
</project>

```

File: oauth2/custom-token-issuer/src/main/java/org/wso2/carbon/identity/extensions/CustomJWTTokenIssuer.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			if (log.isDebugEnabled()) {
+				log.debug("Scope not found for the jwt acccess token");
+			}
+			JWTClaimsSet.Builder jwtClaimsSetBuilder = new JWTClaimsSet.Builder(jwtClaimsSet);
+			return jwtClaimsSetBuilder.build();
+
+		}
+	}
+
+}
```
Comment (mefarazath) [inline oauth2/custom-token-issuer/src/main/java/org/wso2/carbon/identity/extensions/CustomJWTTokenIssuer.java:69]:
```suggestion
}

```

File: oauth2/custom-token-issuer/src/README.MD [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  "iss": "https://localhost:9443/oauth2/token",
+  "exp": 1606306530,
+  "iat": 1606302930,
+  "jti": "540449c8-c6f4-4488-9c3c-e1bbab2aeff0"
+}
+
+
+
+
+
```
Comment (mefarazath) [inline oauth2/custom-token-issuer/src/README.MD:70]: Remove unnecessary new lines

File: oauth2/custom-token-issuer/src/README.MD [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+Replace org.wso2.carbon.identity.extensions. CustomJWTTokenIssuer with your custom class name.
+
+Restart the WSO2 Identity Server.
+
+Command use to generate JWT
+
+curl -v -X POST -H "Authorization: Basic d3o3RFR0RFBHOTZiZmZBdUlIRTdPRUpKN3RVYTpKNmZkaHAwWTNyTDlVSzhJME1xYVMxTWZpekVh" -k -d "grant_type=password&username=<username>&password=<password>" -H "Content-Type:application/x-www-form-urlencoded" "https://<HOST-IP>:<HOST-PORT>/oauth2/token?scope=customerProfile.read%20customerProfile.write%20customerProfile.read%20policyAdmin.read%20healthCareProviders.read%20healthCareProviders.write%20customerServices.write%20customerServices.read%20notifications.read%20claims.read%20claims.write"
+
+
+
```
Comment (piraveena) [inline oauth2/custom-token-issuer/src/README.MD:33]: remove new line

File: oauth2/custom-token-issuer/src/main/java/org/wso2/carbon/identity/extensions/CustomJWTTokenIssuer.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	private static final String SCOPE = "scope";
+	String scope;
+	String[] scopeSplit;
+
+	public CustomJWTTokenIssuer() throws IdentityOAuth2Exception {
+		super();
+		if (log.isDebugEnabled()) {
+			log.debug("CustomJWT Access token Issuer is initiated");
+		}
+
```
Comment (piraveena) [inline oauth2/custom-token-issuer/src/main/java/org/wso2/carbon/identity/extensions/CustomJWTTokenIssuer.java:40]: remove new line.

File: oauth2/custom-token-issuer/src/main/java/org/wso2/carbon/identity/extensions/CustomJWTTokenIssuer.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import com.nimbusds.jwt.JWTClaimsSet;
+
+public class CustomJWTTokenIssuer extends JWTTokenIssuer {
+	private static final Log log = LogFactory.getLog(CustomJWTTokenIssuer.class);
+	private static final String SCOPE = "scope";
+	String scope;
+	String[] scopeSplit;
+
+	public CustomJWTTokenIssuer() throws IdentityOAuth2Exception {
```
Comment (piraveena) [inline oauth2/custom-token-issuer/src/main/java/org/wso2/carbon/identity/extensions/CustomJWTTokenIssuer.java:35]: new line should be there after the method signature. Apply wso2 code style https://sites.google.com/a/wso2.com/engineering/Home/masteringintellijidea

File: oauth2/custom-token-issuer/src/main/java/org/wso2/carbon/identity/extensions/CustomJWTTokenIssuer.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+			JWTClaimsSet.Builder jwtClaimsSetBuilder = new JWTClaimsSet.Builder(jwtClaimsSet);
+			jwtClaimsSetBuilder.claim(SCOPE, scopeSplit);
+			return jwtClaimsSetBuilder.build();
+		} else {
+			if (log.isDebugEnabled()) {
+				log.debug("Scope not found for the jwt acccess token");
+			}
+			JWTClaimsSet.Builder jwtClaimsSetBuilder = new JWTClaimsSet.Builder(jwtClaimsSet);
+			return jwtClaimsSetBuilder.build();
+
```
Comment (piraveena) [inline oauth2/custom-token-issuer/src/main/java/org/wso2/carbon/identity/extensions/CustomJWTTokenIssuer.java:65]: remove new line

**Review by piraveena — CHANGES_REQUESTED** [review]

**Comment by isharak** [comment]
@RajeshTammali 
Since our policy is not to keep PR open for a long time I'll close this PR.
Can you please address the change requests and resend the PR again.

Thanks.

**Comment by mefarazath** [comment]
Improved PR added with https://github.com/wso2/samples-is/pull/275

**Pull request closed without merging**